### PR TITLE
xds: Ensures the xDS cluster is initialized before starting ADS mux

### PR DESCRIPTION
**Pull request by abeyad**
This commit addresses an issue outlined in
https://github.com/envoyproxy/envoy/issues/27702 where, when configuring xDS to use an ADS gRPC mux connecting to an Envoy cluster (via EnvoyGrpc), if the cluster requires asynchronous initialization (e.g. a DNS cluster that must asynchronously resolve the endpoints before it can be used), then the ADS mux attempted to initialize before the cluster was ready.  This caused the underlying GrpcStream to fail to establish the connection and set off a retry timer for [0,1s).

Instead, this commit ensures that we don't try initializing the ADS mux and connecting to the xDS backend until the Envoy cluster for the xDS backend is fully initiliazed and ready.

As part of testing this change, we had to move some of the ClusterManagerImpl initialization functionality to a separate init() method, instead of running it all in the constructor. Running all of the ClusterManagerImpl's lifecycle from the constructor prevents derived classes from overriding any of the functionality (e.g. to override certain functions in tests), because when running in the base class's constructor, only base class methods are executed, not the derived class's.

It turns out that several tests in `cluster_manager_impl_test.cc` depended on a derived class of ClusterManagerImpl which offered mocked behavior, but the overridden methods in the derived class were actually never getting invoked (because they were called from the base ClusterManagerImpl's constructor). We fix those tests in this commit too.

Risk Level: Low
Testing: unit test
Docs Changes: N/A
Release Notes: Fixed a bug (https://github.com/envoyproxy/envoy/issues/27702) that caused ADS initialization
    to fail on the first attempt and set a back-off retry interval of up to 1 second, if ADS is
    using an Envoy Cluster for the backend. The issue was fixed to ensure that ADS initialization
    happens after the Envoy Cluster it depends upon has been properly initialized. ADS that does
    not depend on an Envoy Cluster (i.e. GoogleGrpc) is not affected by this change.
Platform Specific Features: N/A

### CODE REVIEW

**Comment by abeyad** [comment]
@adisuissa this PR is ready for review

File: source/common/upstream/cluster_manager_impl.h
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.h:257]: please add to the comment that this may throw exceptions.

File: source/common/upstream/cluster_manager_impl.cc
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.cc:2187]: nit: I don't think a call to `new ClusterManagerImpl` is needed here.

File: source/common/upstream/cluster_manager_impl.cc
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.cc:93]: Does this need to address google_grpc as well?
If not, please add a comment justifying it.

File: source/common/upstream/cluster_manager_impl.h
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.h:137]: nit: const

File: source/common/upstream/cluster_manager_impl.h
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.h:886]: I should probably know this, but is the bootstrap object lifetime longer than the cluster-manager initialization? (I'm worried about non `init()` methods using this bootstrap).
Consider passing the bootstrap as a parameter to the `init()` method.

File: source/common/upstream/cluster_manager_impl.h
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.h:935]: Why does this need to be atomic?

File: test/common/upstream/test_cluster_manager.h
Comment (adisuissa) [inline test/common/upstream/test_cluster_manager.h:180]: Can the `clusterManagerFromProto()` be used in this case?

File: source/common/upstream/cluster_manager_impl.cc
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.cc:1295]: I think `postThreadLocalClusterUpdate` can be called more than once. In that case, IIUC, `ads_mux_->start()` may be initiated more than once, no?
I suggest also adding an assertion on the `start()` that it is only called once, and consider adding an integration test that updates the endpoints of the ADS cluster.

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM. Left a few comments.
Although this is not strictly viewable from a user-perspective, I think a release-note should be added, as the change may delay when Envoy connects to the ADS server.

**Comment by abeyad** [comment]
> Thanks, overall LGTM. Left a few comments. Although this is not strictly viewable from a user-perspective, I think a release-note should be added, as the change may delay when Envoy connects to the ADS server.

Thanks for the feedback @adisuissa !  I added a release note too, although in this case, I think the change is a strict improvement, since it shouldn't affect GoogleGrpc clusters, and it triggers ADS mux initialization immediately after the ADS Envoy cluster becomes available, so it shouldn't cause a delay in ADS initialization and should prevent back-off retry delays for ADS clusters that require async initialization (e.g. to resolve DNS hosts).

File: source/common/upstream/cluster_manager_impl.cc
Comment (adisuissa) [reply source/common/upstream/cluster_manager_impl.cc:2191]: Ack, thanks.

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM.
Left a minor comment.
Not sure if this needs to be runtime guarded or not.
Also consider merging this after the Envoy version is released (probably the upcoming day or two).

/assign-from @envoyproxy/senior-maintainers

**Comment by abeyad** [comment]
> Thanks, overall LGTM.
> Left a minor comment.
> Not sure if this needs to be runtime guarded or not.
> Also consider merging this after the Envoy version is released (probably the upcoming day or two).

Thanks for the review, @adisuissa !  Happy to wait until after the Envoy release.  In terms of runtime guard, I considered that initially but didn't manage to convince myself it was needed, but if you or Harvey feel differently, happy to add it.

File: source/common/upstream/cluster_manager_impl.cc
Comment (htuch) [inline source/common/upstream/cluster_manager_impl.cc:93]: Fair, but then maybe rename the method to something more specific, e.g. `isBlockingAdsCluster` or whatever you think best.

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_impl.cc:96]: How do we get into this situation now (where it wasn't possible before)?

File: source/common/upstream/cluster_manager_impl.cc
Comment (htuch) [inline source/common/upstream/cluster_manager_impl.cc:404]: One alternative here is to give each cluster a post-load/initialization callback which we then pass the ADS start continuation to. I don't think it's strictly necessary but throwing it out there in case there are other use cases I'm missing.

**Review by htuch — COMMENTED** [review]
LGTM modulo comments. Thanks.

File: source/common/upstream/cluster_manager_impl.cc
Comment (abeyad) [reply source/common/upstream/cluster_manager_impl.cc:407]: This is an interesting idea and thought about it myself, but then wasn't really sure if it was worth adding the extra logic in as I didn't see any other use cases.  I think I'll add a note that if we end up finding we need to do more of this "post cluster init" stuff, it would be good to build the ClusterManagerImpl interfaces to support it.  WDYT?

**Comment by abeyad** [comment]
> LGTM modulo comments. Thanks.

Thanks @htuch !  It's ready for another round of review

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (htuch) [reply source/extensions/config_subscription/grpc/grpc_mux_impl.cc:96]: OK, thanks. Normally not a super fan of code that can't be unit tested (as per the assertion), but in this case it's really trivial and essentially a nop.

File: source/common/upstream/cluster_manager_impl.cc
Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc:407]: Yes, that's fine, this was just a suggestion, fine to take a note on this.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by abeyad** [comment]
thanks for the review @htuch !

**Pull request merged**
